### PR TITLE
Add observation_aggregator

### PR DESCRIPTION
**Pull request by shmurai**
This commit adds `chainermn.extensions.observation_aggregator` which summarizes a specific observation in the trainer over all processes.
This is the implementation of #6974.

### CODE REVIEW

**Comment by shmurai** [comment]
Currently, aggregation (and corresponding communication) occurs every time when the observation is registered.
To avoid worsening the performance, it will be better to decrease the frequency of communication.

**Comment by shmurai** [comment]
I changed the structure of aggregator so that the frequency of communication can be controlled.
Also, I added MultiNodeEarlyStoppingTrigger.

**Comment by keisukefukuda** [comment]
pfnci, test this please

File: chainermn/extensions/multi_node_early_stopping_trigger.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+def _random_suffix(n):
+    return ''.join([random.choice(string.ascii_letters) for i in range(n)])
+
+
+class MultiNodeEarlyStoppingTrigger(object):
+
+    def __init__(self, comm, check_trigger=(1, 'epoch'), monitor='main/loss',
+                 patience=None, mode='auto', verbose=False,
+                 max_trigger=(100, 'epoch'), **kwargs):
+
```
Comment (shu65) [inline chainermn/extensions/multi_node_early_stopping_trigger.py]: Could you add the description about `MultiNodeEarlyStoppingTrigger` ?
ref:
https://docs.chainer.org/en/stable/reference/generated/chainer.training.triggers.EarlyStoppingTrigger.html

Comment (shmurai) [reply chainermn/extensions/multi_node_early_stopping_trigger.py]: fixed.

File: tests/chainermn_tests/extensions_tests/test_multi_node_early_stopping_trigger.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        trainer.observation = {key: accuracy}
+        self.assertEqual(trigger(trainer), expected)
+
+
+class TestMultiNodeEarlyStoppingTrigger(unittest.TestCase):
+
+    def setUp(self):
+        self.communicator = chainermn.create_communicator('naive')
+
+    def test_early_stopping_trigger_with_accuracy(self):
```
Comment (shu65) [inline tests/chainermn_tests/extensions_tests/test_multi_node_early_stopping_trigger.py]: Could you add tests with cupy version?

File: tests/chainermn_tests/extensions_tests/test_observation_aggregator.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+class DummyChain(chainer.Chain):
+
+    def __init__(self):
+        super(DummyChain, self).__init__()
+
+    def forward(self, x):
+        return chainer.Variable(x, grad=np.array([0]))
+
+
+class TestObservationAggregator(unittest.TestCase):
```
Comment (shu65) [inline tests/chainermn_tests/extensions_tests/test_observation_aggregator.py]: Could you add tests with cupy version?

Comment (shmurai) [reply tests/chainermn_tests/extensions_tests/test_observation_aggregator.py]: added.

File: tests/chainermn_tests/extensions_tests/test_multi_node_early_stopping_trigger.py (hunk 2 above)
Comment (shmurai) [reply tests/chainermn_tests/extensions_tests/test_multi_node_early_stopping_trigger.py]: added.

**Comment by shu65** [comment]
pfnci, test this please

**Pull request merged**